include/libcyphal/transport/can/svc_rx_sessions.hpp

```diff
@@ -0,0 +1,169 @@
+/// Not supposed to be used directly by the users of the library.
+///
+namespace detail
+{
+
+/// @brief A template class to represent a service request/response RX session (both for server and client sides).
+///
+/// @tparam Interface_ Type of the session interface.
+///                    Could be either `IRequestRxSession` or `IResponseRxSession`.
+/// @tparam Params Type of the session parameters.
+///                Could be either `RequestRxParams` or `ResponseRxParams`.
+/// @tparam TransferKind Kind of the service transfer.
+///                      Could be either `CanardTransferKindRequest` or `CanardTransferKindResponse`.
+///
+template <typename Interface_, typename Params, CanardTransferKind TransferKind>
+class SvcRxSession final : public Interface_, private IRxSessionDelegate
+{
+    /// @brief Defines specification for making interface unique ptr.
+    ///
+    struct Spec
+    {
+        using Interface = Interface_;
+        using Concrete  = SvcRxSession;
+
+        // In use to disable public construction.
+        // See https://seanmiddleditch.github.io/enabling-make-unique-with-private-constructors/
+        explicit Spec() = default;
+    };
+
+public:
+    CETL_NODISCARD static Expected<UniquePtr<Interface_>, AnyError> make(TransportDelegate& delegate,
+                                                                         const Params&      params)
+    {
+        if (params.service_id > CANARD_SERVICE_ID_MAX)
+        {
+            return ArgumentError{};
+        }
+
+        auto session = libcyphal::detail::makeUniquePtr<Spec>(delegate.memory(), Spec{}, delegate, params);
+        if (session == nullptr)
+        {
+            return MemoryError{};
+        }
+
+        return session;
+    }
+
+    SvcRxSession(Spec, TransportDelegate& delegate, const Params& params)
+        : delegate_{delegate}
+        , params_{params}
+        , subscription_{}
+        , last_rx_transfer_{}
+    {
+        const int8_t result = ::canardRxSubscribe(&delegate.canard_instance(),
+                                                  TransferKind,
+                                                  static_cast<CanardPortID>(params_.service_id),
+                                                  static_cast<std::size_t>(params_.extent_bytes),
+                                                  CANARD_DEFAULT_TRANSFER_ID_TIMEOUT_USEC,
+                                                  &subscription_);
+        (void) result;
+        CETL_DEBUG_ASSERT(result >= 0, "There is no way currently to get an error here.");
+        CETL_DEBUG_ASSERT(result > 0, "New subscription supposed to be made.");
+
+        subscription_.user_reference = static_cast<IRxSessionDelegate*>(this);
+    }
+
+    ~SvcRxSession() final
+    {
+        ::canardRxUnsubscribe(&delegate_.canard_instance(),
+                              TransferKind,
+                              static_cast<CanardPortID>(params_.service_id));
```

Comment:
No path fail as well, but if you think its worth to assert I can do it (here and for msg rx session).

Reply:
I like the debug asserts as long as we are able to verify them.